pkg/reconciler/v1alpha1/clusteringress/resources/virtual_service.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	istiov1alpha1 "github.com/knative/pkg/apis/istio/common/v1alpha1"
+	"github.com/knative/pkg/apis/istio/v1alpha3"
+	"github.com/knative/pkg/kmeta"
+	"github.com/knative/serving/pkg/apis/networking"
+	"github.com/knative/serving/pkg/apis/networking/v1alpha1"
+	"github.com/knative/serving/pkg/apis/serving"
+	"github.com/knative/serving/pkg/reconciler"
+	"github.com/knative/serving/pkg/reconciler/v1alpha1/clusteringress/resources/names"
+	"github.com/knative/serving/pkg/system"
+)
+
+// MakeVirtualService creates an Istio VirtualService as network programming.
+// Such VirtualService specifies which Gateways and Hosts that it applies to,
+// as well as the routing rules.
+func MakeVirtualService(ci *v1alpha1.ClusterIngress) *v1alpha3.VirtualService {
+	vs := &v1alpha3.VirtualService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            names.VirtualService(ci),
+			Namespace:       system.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*kmeta.NewControllerRef(ci)},
+		},
+		Spec: *makeVirtualServiceSpec(ci),
+	}
+
+	// Populate the ClusterIngress labels.
+	if vs.Labels == nil {
+		vs.Labels = make(map[string]string)
+	}
+	vs.Labels[networking.IngressLabelKey] = ci.Name
```

Comment:
If we decide to make it with ` EnqueueControllerOf`, then the  ingress label can be removed.
As for the route label, it is indeed not used in the code, but I think it can be useful for users to do query/debugging: Users might care about which route a virtualservice belongs to, but the relationship is not so obvious.

Reply:
I see. Thanks for the explanation!